File: Submissions/865_Smallest_Subtree_with_all_the_Deepest_Nodes.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    TreeNode* subtreeWithAllDeepest(TreeNode* root) {
        auto result = recursive(root);
        return result.second;
    }
    
    pair<int, TreeNode*> recursive(TreeNode* node){
        if(!node) return std::make_pair(0, nullptr);
        
        auto left_result = recursive(node->left);
        auto right_result = recursive(node->right);
        
        if (left_result.first == right_result.first) return std::make_pair(left_result.first+1, node);
        else if (left_result.first > right_result.first) return std::make_pair(left_result.first+1, left_result.second);
        else return std::make_pair(right_result.first+1, right_result.second);        
    }
};

File: Kotlin/865_Smallest_Subtree_with_all_the_Deepest_Nodes.kt
/**
 * Example:
 * var ti = TreeNode(5)
 * var v = ti.`val`
 * Definition for a binary tree node.
 * class TreeNode(var `val`: Int) {
 *     var left: TreeNode? = null
 *     var right: TreeNode? = null
 * }
 */
class Solution {
    fun subtreeWithAllDeepest(root: TreeNode?): TreeNode? {
        val pair = recursive(root)
        return pair.second
    }
    
    fun recursive(node: TreeNode?): Pair<Int, TreeNode?> {
        if(node == null) return Pair(0, null)
        
        val left = recursive(node.left)
        val right = recursive(node.right)
        
        if(left.first == right.first) return Pair(left.first+1, node)
        else if(left.first > right.first) return Pair(left.first+1, left.second)
        else return Pair(right.first+1, right.second)
    }
}